contracts/contracts/TACoApplication.sol

```diff
@@ -957,4 +1051,39 @@ contract TACoApplication is
         stakingProviderWrapper[0] = _stakingProvider;
         tStaking.seize(_penalty, 100, _investigator, stakingProviderWrapper);
     }
+
+    /**
+     * @notice Penalize the staking provider's future reward
+     * @param _stakingProvider Staking provider address
+     */
+    function penalize(address _stakingProvider) external updateReward(_stakingProvider) {
+        require(
+            msg.sender == address(childApplication),
+            "Only child application allowed to penalize"
+        );
+
+        if (_stakingProvider == address(0)) {
+            return;
+        }
+
+        StakingProviderInfo storage info = stakingProviderInfo[_stakingProvider];
+        uint96 before = effectiveAuthorized(info.authorized, info.penaltyPercent);
+        info.endPenalty = uint64(block.timestamp + penaltyDuration);
+        info.penaltyPercent = penaltyDefault;
```

Comment:
Can we add logic to this function such that the `penaltyPercent` is increased by previous violations? To make it as simple as possible:  
- `penaltyDuration` can be a constant (e.g. 1 month for all violations) 
- A violation that occurs during an already penalized reward period simply restarts the 1 month clock from the day that it occurs, but with increased severity. It does not lead to any special outcomes (e.g. summing `penaltyPercent`s) 
- An address's violation count resets after a universal fixed duration (e.g. 24 months), after which point a new violation would result in the least severe `penaltyPercent`. 

Hypothetical parameters: the first incomplete DKG violation leads to a 25% reduction in rewards for a month. The second leads to a 50%, third to a 75%, and fourth (& fifth, sixth...nth) to a 100% reduction. 

Depending on resultant behavior, we can consider increasing `penaltyDuration` for violations beyond the fourth. The rationale for avoiding this in the first version is that we don't want extended periods where nodes are earning nothing, as that may encourage them to exit the network.

Reply:
This sounds like a good idea. Just one consideration:

- About the penalty duration increasing beyond the fourth: I think we should consider setting its max value to the deauthorization delay (183 days) or less. If this reset duration is much higher than the deauth delay, the [reoffending] operators with 100% of the penalty will be incentivized to just remove the operator from the network and set up a new one with a different staking provider address to avoid the penalty. And this can lead to an unstable network with cohorts with unreachable nodes.

Reply:
@arjunhassard I'm not following completely, can you elaborate a bit further?

Reply:
@arjunhassard maybe open this as an issue and I'll address that in separate PR?

Reply:
https://github.com/nucypher/nucypher-contracts/issues/259